source/fe/fe.cc

```diff
@@ -679,18 +679,16 @@ FiniteElement<dim, spacedim>::adjust_quad_dof_index_for_face_orientation(
   AssertIndexRange(index, this->n_dofs_per_quad(face));
   Assert(adjust_quad_dof_index_for_face_orientation_table
              [this->n_unique_quads() == 1 ? 0 : face]
-               .n_elements() == (this->reference_cell().face_reference_cell(
-                                   face) == ReferenceCells::Quadrilateral ?
-                                   8 :
-                                   6) *
-                                  this->n_dofs_per_quad(face),
+               .n_elements() ==
+           (this->reference_cell().n_face_orientations(face)) *
+             this->n_dofs_per_quad(face),
          ExcInternalError());
   return index +
          adjust_quad_dof_index_for_face_orientation_table
            [this->n_unique_quads() == 1 ? 0 : face](index,
-                                                    (face_orientation ? 4 : 0) +
-                                                      (face_flip ? 2 : 0) +
-                                                      (face_rotation ? 1 : 0));
+                                                    (face_orientation ? 1 : 0) +
+                                                      (face_rotation ? 2 : 0) +
+                                                      (face_flip ? 4 : 0));
```

Comment:
This is apparently not the same encoding used in other places in this class -- judging by the fact that a number of tests fail. You're completely correct that we ought to not do the en/decoding of these flags by hand, but using the corresponding functions in `ReferenceCell`. Maybe the right step would be to just search through the entire file for `face_flip` or some such, and replace all of these hand encodings by what `ReferenceCell` offers.

Reply:
Definitely - this would be a good time to simultaneously make our encodings uniform and do the follow-ups listed in #12878. We have at least three different orientation conventions in the library and I would much rather have one.

Reply:
Yes, three is too many when there are really only 6 possible (reasonable) encodings of three bits of information...